File: completions-streaming/final/python/auth/token_manager.py
"""
Token manager for Gloo AI OAuth2 authentication.

Handles client credentials grant type with automatic token refresh.
"""

import os
import requests
from datetime import datetime, timedelta

# API Endpoints
TOKEN_URL = "https://platform.ai.gloo.com/oauth2/token"

# Token state (module-level for simplicity)
access_token = None
token_expiry = None


def get_access_token() -> dict:
    """
    Retrieve an OAuth2 access token from Gloo AI.

    Uses the client credentials grant type with GLOO_CLIENT_ID and
    GLOO_CLIENT_SECRET from environment variables.

    Returns:
        dict: Token response containing access_token and expires_in

    Raises:
        ValueError: If credentials are not set
        Exception: If the token request fails
    """
    client_id = os.getenv("GLOO_CLIENT_ID")
    client_secret = os.getenv("GLOO_CLIENT_SECRET")

    if not client_id or not client_secret:
        raise ValueError(
            "Missing credentials. Set GLOO_CLIENT_ID and GLOO_CLIENT_SECRET "
            "environment variables."
        )

    payload = {
        "grant_type": "client_credentials",
        "client_id": client_id,
        "client_secret": client_secret,
    }

    try:
        response = requests.post(TOKEN_URL, data=payload)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        raise Exception(f"Failed to get access token: {str(e)}")


def ensure_valid_token() -> str:
    """
    Ensure we have a valid access token, refreshing if necessary.

    Caches the token at module level and refreshes it 5 minutes before
    expiry to avoid using an expired token mid-request.

    Returns:
        str: Valid access token
    """
    global access_token, token_expiry

    # Check if we need a new token
    if not access_token or not token_expiry or datetime.now() >= token_expiry:
        token_data = get_access_token()
        access_token = token_data["access_token"]
        # Set expiry with 5-minute buffer
        expires_in = token_data.get("expires_in", 3600)
        token_expiry = datetime.now() + timedelta(seconds=expires_in - 300)

    return access_token

File: completions-streaming/final/typescript/src/auth/tokenManager.ts
/**
 * Token manager for Gloo AI OAuth2 authentication.
 *
 * Handles client credentials grant type with automatic token refresh.
 * Caches the token in module scope and refreshes it 5 minutes before expiry.
 */

import type { TokenInfo } from "../types.js";

const TOKEN_URL = "https://platform.ai.gloo.com/oauth2/token";

let accessToken: string | null = null;
let tokenExpiry: number | null = null;

/**
 * Retrieve an OAuth2 access token from Gloo AI.
 *
 * @returns Promise resolving to a TokenInfo object
 * @throws Error if credentials are missing or the request fails
 */
export async function getAccessToken(): Promise<TokenInfo> {
  const clientId = process.env.GLOO_CLIENT_ID;
  const clientSecret = process.env.GLOO_CLIENT_SECRET;

  if (!clientId || !clientSecret) {
    throw new Error(
      "Missing credentials. Set GLOO_CLIENT_ID and GLOO_CLIENT_SECRET environment variables."
    );
  }

  const params = new URLSearchParams({
    grant_type: "client_credentials",
    client_id: clientId,
    client_secret: clientSecret,
  });

  const response = await fetch(TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: params.toString(),
  });

  if (!response.ok) {
    throw new Error(`Failed to get access token: HTTP ${response.status}`);
  }

  return response.json() as Promise<TokenInfo>;
}

/**
 * Ensure we have a valid access token, refreshing if necessary.
 *
 * @returns Promise resolving to a valid access token string
 */
export async function ensureValidToken(): Promise<string> {
  const now = Date.now();

  if (!accessToken || !tokenExpiry || now >= tokenExpiry) {
    const tokenData = await getAccessToken();
    accessToken = tokenData.access_token;
    // Set expiry with 5-minute buffer (expires_in is in seconds)
    const expiresIn = tokenData.expires_in ?? 3600;
    tokenExpiry = now + (expiresIn - 300) * 1000;
  }

  return accessToken;
}
